### PR TITLE
Lookup relations - Search external sources

### PR DESCRIPTION
**Pull request by Atmire-Kristof**
This PR is dependent on #531 and expands on the relation lookup window added for submission.

Using the REST API's [external sources endpoints](https://github.com/DSpace/Rest7Contract/blob/master/external-authority-sources.md), it adds a tab for each of the external sources.
These tabs display a list of external entries for the selected source. The user is able to select one or more of these entries to add them to the current selection.

All of the tabs also show the amount of results available for the current query (which is shared between all tabs), even when the relevant tabs are not active.

Currently, the user can only add external entries to the selection. Importing them into the repository as entities/authorities will be part of a separate PR.

### CODE REVIEW

**Comment by benbosman** [comment]
I've created https://github.com/atmire/dspace-angular/pull/3 to clearly display what's new in this branch vs https://github.com/DSpace/dspace-angular/pull/531

File: src/app/core/data/external-source.service.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import { PaginatedSearchOptions } from '../../shared/search/paginated-search-options.model';
+import { hasValue, isNotEmptyOperator } from '../../shared/empty.util';
+import { configureRequest } from '../shared/operators';
+import { RemoteData } from './remote-data';
+import { PaginatedList } from './paginated-list';
+import { ExternalSourceEntry } from '../shared/external-source-entry.model';
+import { DefaultChangeAnalyzer } from './default-change-analyzer.service';
+
+@Injectable()
+export class ExternalSourceService extends DataService<ExternalSource> {
```
Comment (paulo-graca) [inline src/app/core/data/external-source.service.ts]: can you please add type docs?

File: src/app/core/data/external-source.service.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    protected store: Store<CoreState>,
+    protected objectCache: ObjectCacheService,
+    protected halService: HALEndpointService,
+    protected notificationsService: NotificationsService,
+    protected http: HttpClient,
+    protected comparator: DefaultChangeAnalyzer<ExternalSource>) {
+    super();
+  }
+
+  getBrowseEndpoint(options: FindListOptions = {}, linkPath: string = this.linkPath): Observable<string> {
```
Comment (paulo-graca) [inline src/app/core/data/external-source.service.ts]: can you please add type docs?

File: src/app/core/data/relationship.service.ts [hunk 3]
```diff
@@ -254,20 +229,113 @@ export class RelationshipService extends DataService<Relationship> {
...
+
+  private isItemMatchWithItemRD(itemRD$: Observable<RemoteData<Item>>, itemCheck: Item): Observable<boolean> {
+    return itemRD$.pipe(
+      getSucceededRemoteData(),
+      map((itemRD: RemoteData<Item>) => itemRD.payload),
+      map((item: Item) => item.uuid === itemCheck.uuid)
+    );
+  }
+
+  public setNameVariant(listID: string, itemID: string, nameVariant: string) {
```
Comment (paulo-graca) [inline src/app/core/data/relationship.service.ts]: can you please add type docs?

File: src/app/core/data/relationship.service.ts [hunk 4]
```diff
@@ -254,20 +229,113 @@ export class RelationshipService extends DataService<Relationship> {
...
-      this.objectCache.remove(rightItem.payload.self);
-      this.requestService.removeByHrefSubstring(leftItem.payload.self);
-      this.requestService.removeByHrefSubstring(rightItem.payload.self);
-    });
+      map((itemRD: RemoteData<Item>) => itemRD.payload),
+      map((item: Item) => uuids.includes(item.uuid))
+    );
+  }
+
+  getRelationshipByItemsAndLabel(item1: Item, item2: Item, label: string): Observable<Relationship> {
```
Comment (paulo-graca) [inline src/app/core/data/relationship.service.ts]: can you please add type docs?

File: src/app/core/data/relationship.service.ts [hunk 5]
```diff
@@ -254,20 +229,113 @@ export class RelationshipService extends DataService<Relationship> {
...
+      map((itemRD: RemoteData<Item>) => itemRD.payload),
+      map((item: Item) => item.uuid === itemCheck.uuid)
+    );
+  }
+
+  public setNameVariant(listID: string, itemID: string, nameVariant: string) {
+    this.appStore.dispatch(new SetNameVariantAction(listID, itemID, nameVariant));
+  }
+
+  public getNameVariant(listID: string, itemID: string): Observable<string> {
```
Comment (paulo-graca) [inline src/app/core/data/relationship.service.ts]: can you please add type docs?

File: src/app/core/data/relationship.service.ts [hunk 6]
```diff
@@ -254,20 +229,113 @@ export class RelationshipService extends DataService<Relationship> {
...
+    return this.appStore.pipe(
+      select(relationshipStateSelector(listID, itemID))
+    );
+  }
+
+  public removeNameVariant(listID: string, itemID: string) {
```
Comment (paulo-graca) [inline src/app/core/data/relationship.service.ts]: can you please add type docs?

File: src/app/core/data/relationship.service.ts [hunk 7]
```diff
@@ -254,20 +229,113 @@ export class RelationshipService extends DataService<Relationship> {
...
+    this.appStore.dispatch(new RemoveNameVariantAction(listID, itemID));
+  }
+
+  public getNameVariantsByListID(listID: string) {
```
Comment (paulo-graca) [inline src/app/core/data/relationship.service.ts]: can you please add type docs?

File: src/app/core/data/relationship.service.ts [hunk 8]
```diff
@@ -254,20 +229,113 @@ export class RelationshipService extends DataService<Relationship> {
...
+    return this.appStore.pipe(select(relationshipListStateSelector(listID)));
+  }
+
+  public updateNameVariant(item1: Item, item2: Item, relationshipLabel: string, nameVariant: string): Observable<RemoteData<Relationship>> {
```
Comment (paulo-graca) [inline src/app/core/data/relationship.service.ts]: can you please add type docs?

File: src/app/core/services/route.service.ts [hunk 9]
```diff
@@ -183,4 +186,23 @@ export class RouteService {
       map((history: string[]) => history[history.length - 2] || '')
     );
   }
+
+  public addParameter(key, value) {
```
Comment (paulo-graca) [inline src/app/core/services/route.service.ts]: can you please add type docs?

File: src/app/core/services/route.service.ts [hunk 10]
```diff
@@ -183,4 +186,23 @@ export class RouteService {
...
+    this.store.dispatch(new AddParameterAction(key, value));
+  }
+
+  public setParameter(key, value) {
+    this.store.dispatch(new SetParameterAction(key, value));
+  }
+
+  public setCurrentRouteInfo() {
```
Comment (paulo-graca) [inline src/app/core/services/route.service.ts]: can you please add type docs?

File: src/app/core/services/route.service.ts [hunk 11]
```diff
@@ -183,4 +186,23 @@ export class RouteService {
       map((history: string[]) => history[history.length - 2] || '')
     );
   }
+
+  public addParameter(key, value) {
+    this.store.dispatch(new AddParameterAction(key, value));
+  }
+
+  public setParameter(key, value) {
```
Comment (paulo-graca) [inline src/app/core/services/route.service.ts]: can you please add type docs?

File: src/app/core/utilities/equatable.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    }
+    if (typeof object1[key] === 'object') {
+      return !equalsByFields(object1[key], object2[key], Object.keys(object1))
+    }
+    return object1[key] !== object2[key];
+  });
+  return hasNoValue(unequalProperty);
+}
+
+export abstract class EquatableObject<T> {
```
Comment (paulo-graca) [inline src/app/core/utilities/equatable.ts]: can you please add type docs?

File: src/app/entity-groups/research-entities/submission/item-list-elements/external-source-entry/external-source-entry-list-submission-element.component.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+import { Metadata } from '../../../../../core/shared/metadata.utils';
+import { MetadataValue } from '../../../../../core/shared/metadata.models';
+
+@listableObjectComponent(ExternalSourceEntry, ViewMode.ListElement, Context.SubmissionModal)
+@Component({
+  selector: 'ds-external-source-entry-list-submission-element',
+  styleUrls: ['./external-source-entry-list-submission-element.component.scss'],
+  templateUrl: './external-source-entry-list-submission-element.component.html'
+})
+export class ExternalSourceEntryListSubmissionElementComponent extends AbstractListableElementComponent<ExternalSourceEntry> implements OnInit {
```
Comment (paulo-graca) [inline src/app/entity-groups/research-entities/submission/item-list-elements/external-source-entry/external-source-entry-list-submission-element.component.ts]: can you please add type docs?

File: src/app/entity-groups/research-entities/submission/name-variant-modal/name-variant-modal.component.ts [hunk 14]
```diff
@@ -0,0 +1,14 @@
+import { Component, Input } from '@angular/core';
+import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
+
+@Component({
+  selector: 'ds-name-variant-modal',
+  templateUrl: './name-variant-modal.component.html',
+  styleUrls: ['./name-variant-modal.component.scss']
+})
+export class NameVariantModalComponent {
```
Comment (paulo-graca) [inline src/app/entity-groups/research-entities/submission/name-variant-modal/name-variant-modal.component.ts]: can you please add type docs?

File: src/app/shared/form/builder/ds-dynamic-form-ui/models/disabled/dynamic-disabled.component.ts [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+import { DynamicFormControlComponent, DynamicFormLayoutService, DynamicFormValidationService } from '@ng-dynamic-forms/core';
+import { FormGroup } from '@angular/forms';
+import { DynamicDisabledModel } from './dynamic-disabled.model';
+import { RelationshipTypeService } from '../../../../../../core/data/relationship-type.service';
+
+@Component({
+  selector: 'ds-dynamic-disabled',
+  templateUrl: './dynamic-disabled.component.html'
+})
+export class DsDynamicDisabledComponent extends DynamicFormControlComponent {
```
Comment (paulo-graca) [inline src/app/shared/form/builder/ds-dynamic-form-ui/models/disabled/dynamic-disabled.component.ts]: can you please add type docs?

File: src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/dynamic-lookup-relation-modal.component.ts [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+  templateUrl: './dynamic-lookup-relation-modal.component.html',
+  providers: [
+    {
+      provide: SEARCH_CONFIG_SERVICE,
+      useClass: SearchConfigurationService
+    }
+  ]
+})
+
+export class DsDynamicLookupRelationModalComponent implements OnInit, OnDestroy {
```
Comment (paulo-graca) [inline src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/dynamic-lookup-relation-modal.component.ts]: can you please add type docs?

File: src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/external-source-tab/dynamic-lookup-relation-external-source-tab.component.ts [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+      useClass: SearchConfigurationService
+    }
+  ],
+  animations: [
+    fadeIn,
+    fadeInOut
+  ]
+})
+
+export class DsDynamicLookupRelationExternalSourceTabComponent implements OnInit {
```
Comment (paulo-graca) [inline src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/external-source-tab/dynamic-lookup-relation-external-source-tab.component.ts]: can you please add type docs?

File: src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/selection-tab/dynamic-lookup-relation-selection-tab.component.ts [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+  templateUrl: './dynamic-lookup-relation-selection-tab.component.html',
+  providers: [
+    {
+      provide: SEARCH_CONFIG_SERVICE,
+      useClass: SearchConfigurationService
+    }
+  ]
+})
+
+export class DsDynamicLookupRelationSelectionTabComponent {
```
Comment (paulo-graca) [inline src/app/shared/form/builder/ds-dynamic-form-ui/relation-lookup-modal/selection-tab/dynamic-lookup-relation-selection-tab.component.ts]: can you please add type docs?

File: src/app/shared/form/builder/models/relationship-options.model.ts [hunk 19]
```diff
@@ -0,0 +1,12 @@
+const RELATION_METADATA_PREFIX = 'relation.'
+
+export class RelationshipOptions {
```
Comment (paulo-graca) [inline src/app/shared/form/builder/models/relationship-options.model.ts]: can you please add type docs?

File: src/app/shared/form/builder/parsers/disabled-field-parser.ts [hunk 20]
```diff
@@ -0,0 +1,11 @@
+import { FieldParser } from './field-parser';
+import { FormFieldMetadataValueObject } from '../models/form-field-metadata-value.model';
+import { DsDynamicDisabledModelConfig, DynamicDisabledModel } from '../ds-dynamic-form-ui/models/disabled/dynamic-disabled.model';
+
+export class DisabledFieldParser extends FieldParser {
```
Comment (paulo-graca) [inline src/app/shared/form/builder/parsers/disabled-field-parser.ts]: can you please add type docs?

**Comment by paulo-graca** [comment]
Thank you for referring the internal diff - https://github.com/atmire/dspace-angular/pull/3/files - it helps the code reviewing process!
Despite the missing type docs, this PR is working as expected. But I'm finding some issues that I think are related with https://github.com/DSpace/dspace-angular/pull/531 and also the questionable search results of ORCID (but unrelated with this PR).
Overall this PR it's ok by me and I will give my thumbs up as soon as it passes the checks and the requested comments are added.

File: src/app/core/cache/models/normalized-external-source-entry.model.ts [hunk 21]
```diff
@@ -0,0 +1,33 @@
+import { autoserialize, autoserializeAs, inheritSerialization } from 'cerialize';
+import { NormalizedObject } from './normalized-object.model';
+import { ExternalSourceEntry } from '../../shared/external-source-entry.model';
+import { mapsTo } from '../builders/build-decorators';
+import { MetadataMap, MetadataMapSerializer } from '../../shared/metadata.models';
+
+@mapsTo(ExternalSourceEntry)
+@inheritSerialization(NormalizedObject)
+export class NormalizedExternalSourceEntry extends NormalizedObject<ExternalSourceEntry> {
```
Comment (tdonohue) [inline src/app/core/cache/models/normalized-external-source-entry.model.ts]: New model class needs Typedocs, please

File: src/app/core/cache/models/normalized-external-source.model.ts [hunk 22]
```diff
@@ -0,0 +1,26 @@
+import { autoserialize, inheritSerialization } from 'cerialize';
+import { NormalizedObject } from './normalized-object.model';
+import { ExternalSource } from '../../shared/external-source.model';
+import { mapsTo } from '../builders/build-decorators';
+
+@mapsTo(ExternalSource)
+@inheritSerialization(NormalizedObject)
+export class NormalizedExternalSource extends NormalizedObject<ExternalSource> {
```
Comment (tdonohue) [inline src/app/core/cache/models/normalized-external-source.model.ts]: Again, class needs Typedocs, please

**Comment by paulo-graca** [comment]
Just want to leave a message that is still working as expected: 
image

**Comment by benbosman** [comment]
I've updated this branch to ensure GIT will only display the commits which differ with master

**Pull request merged**